test/basic_json.cpp

```diff
@@ -0,0 +1,1480 @@
+//-----------------------------------------------------------------------------
+TEST(general, parsing)
+{
+    auto tree = parse("{foo: 1}");
+
+    char cmpbuf[128] = {0};
+    substr cmp(cmpbuf);
+    size_t ret;
+
+    ret = cat(cmp, tree["foo"].val());
+    EXPECT_EQ(cmp.first(ret), "1");
+
+    ret = cat(cmp, tree["foo"].key());
+    EXPECT_EQ(cmp.first(ret), "foo");
+}
+
+TEST(general, emitting)
+{
+    std::string cmpbuf;
+    std::string cmpbuf2;
+
+    Tree tree;
+    auto r = tree.rootref();
+
+    r |= MAP;  // this is needed to make the root a map
+
+    r["foo"] = "1"; // ryml works only with strings.
+    // Note that the tree will be __pointing__ at the
+    // strings "foo" and "1" used here. You need
+    // to make sure they have at least the same
+    // lifetime as the tree.
+
+    auto s = r["seq"]; // does not change the tree until s is written to.
+    s |= SEQ;
+    r["seq"].append_child() = "bar0"; // value of this child is now __pointing__ at "bar0"
+    r["seq"].append_child() = "bar1";
+    r["seq"].append_child() = "bar2";
+
+    //print_tree(tree);
+
+    // emit to stdout (can also emit to FILE* or ryml::span)
+    emitrs_json(tree, &cmpbuf);
+    c4::substr c4cmp(to_substr(cmpbuf));
+    c4::yml::Tree res = c4::yml::parse(c4cmp);
+    emitrs(res, &cmpbuf2);
+    const char* exp = R"(foo: 1
+seq:
+  - bar0
+  - bar1
+  - bar2
+)";
+    EXPECT_EQ(cmpbuf2, exp);
+
+    // serializing: using operator<< instead of operator=
+    // will make the tree serialize the value into a char
+    // arena inside the tree. This arena can be reserved at will.
+    int ch3 = 33, ch4 = 44;
+    s.append_child() << ch3;
+    s.append_child() << ch4;
+
+    {
+        std::string tmp = "child5";
+        s.append_child() << tmp;
+        // now tmp can go safely out of scope, as it was
+        // serialized to the tree's internal string arena
+    }
+
+    emitrs_json(tree, &cmpbuf);
+    c4cmp=to_substr(cmpbuf);
+    res = c4::yml::parse(c4cmp);
+    emitrs(res, &cmpbuf2);
+    exp = R"(foo: 1
+seq:
+  - bar0
+  - bar1
+  - bar2
+  - 33
+  - 44
+  - child5
+)";
+    EXPECT_EQ(cmpbuf2, exp);
+
+    // to serialize keys:
+    int k=66;
+    r.append_child() << key(k) << 7;
+
+    emitrs_json(tree, &cmpbuf);
+    c4cmp=to_substr(cmpbuf);    
+    res = c4::yml::parse(c4cmp);
+    emitrs(res, &cmpbuf2);
+    exp = R"(foo: 1
+seq:
+  - bar0
+  - bar1
+  - bar2
+  - 33
+  - 44
+  - child5
+66: 7
+)";
+    EXPECT_EQ(cmpbuf2, exp);
+}
+
+TEST(general, map_to_root)
+{
+    std::string cmpbuf; const char *exp;
+    std::map<std::string, int> m({{"bar", 2}, {"foo", 1}});
+    Tree t;
+    t.rootref() << m;
+
+    emitrs_json(t, &cmpbuf);
+    exp = "{\"bar\": 2,\"foo\": 1}";
+    EXPECT_EQ(cmpbuf, exp);
+
+    t["foo"] << 10;
+    t["bar"] << 20;
+
+    m.clear();
+    t.rootref() >> m;
+
+    EXPECT_EQ(m["foo"], 10);
+    EXPECT_EQ(m["bar"], 20);
+}
+
+TEST(general, print_tree)
```

Comment:
please remove. this test does not touch any json logic, and is a pure duplicate of the original.

Reply:
done